File: src/main/java/org/jetbrains/fortran/lang/parser/IdentifierParser.java
package org.jetbrains.fortran.lang.parser;

import com.intellij.lang.PsiBuilder;
import com.intellij.lang.parser.GeneratedParserUtilBase;

import static com.intellij.lang.parser.GeneratedParserUtilBase.*;
import static org.jetbrains.fortran.lang.FortranTypes.*;
import static org.jetbrains.fortran.lang.psi.FortranTokenType.KEYWORDS;
import static org.jetbrains.fortran.lang.psi.FortranTokenType.WORD;

public class IdentifierParser implements GeneratedParserUtilBase.Parser {
    @Override
    public boolean parse(final PsiBuilder builder, final int level) {
        if (!recursion_guard_(builder, level, "Identifier")) return false;
        boolean result;
        PsiBuilder.Marker marker = enter_section_(builder);
        result = consumeToken(builder, IDENTIFIER);
        if (!result) {
            if (builder.getTokenType() == WORD || KEYWORDS.contains(builder.getTokenType())) {
                builder.remapCurrentToken(IDENTIFIER);
                result = consumeToken(builder, IDENTIFIER);
            }
        }
        exit_section_(builder, marker, null, result);
        return result;
    }
}

File: src/main/kotlin/org/jetbrains/fortran/lang/parser/IdentifierParser.kt
package org.jetbrains.fortran.lang.parser

import com.intellij.lang.PsiBuilder
import org.jetbrains.fortran.lang.FortranTypes.IDENTIFIER
import org.jetbrains.fortran.lang.parser.FortranParserUtil.consumeToken
import org.jetbrains.fortran.lang.parser.FortranParserUtil.enter_section_
import org.jetbrains.fortran.lang.parser.FortranParserUtil.exit_section_
import org.jetbrains.fortran.lang.parser.FortranParserUtil.recursion_guard_
import org.jetbrains.fortran.lang.psi.FortranTokenSets
import org.jetbrains.fortran.lang.psi.FortranTokenType

class IdentifierParser : FortranParserUtil.Parser {
    override fun parse(builder: PsiBuilder, level: Int): Boolean {
        if (!recursion_guard_(builder, level, "Identifier")) return false
        var result: Boolean
        val marker = enter_section_(builder)
        result = consumeToken(builder, IDENTIFIER)
        if (!result) {
            val tokenType = builder.tokenType
            if (tokenType === FortranTokenType.WORD || FortranTokenSets.KEYWORDS.contains(tokenType)) {
                builder.remapCurrentToken(FortranParserUtil.cloneTTwithBase(tokenType, IDENTIFIER))
                result = consumeToken(builder, IDENTIFIER)
            }
        }
        exit_section_(builder, marker, null, result)
        return result
    }
}
